tests/functional/src/main/java/org/mule/functional/classloading/isolation/classloader/MuleClassLoaderRunnerFactory.java

```diff
@@ -0,0 +1,265 @@
+            if (pluginUrlClassification instanceof ExtensionPluginUrlClassification)
+            {
+                extensionPluginsArtifactClassLoaders.add(pluginCL);
+
+                URL manifestUrl = pluginCL.findResource("META-INF/" + EXTENSION_MANIFEST_FILE_NAME);
+                ExtensionManifest extensionManifest = extensionManager.parseExtensionManifestXml(manifestUrl);
+
+                exportedPackages = extensionManifest.getExportedPackages();
+                exportedResources = extensionManifest.getExportedResources();
+            }
+            else
+            {
+                ClassLoader pluginArtifactClassLoaderToDiscoverModules = new URLClassLoader(pluginUrlClassification.getUrls().toArray(new URL[0]), null);
+                List<MuleModule> modules = withContextClassLoader(pluginArtifactClassLoaderToDiscoverModules, () -> new ClasspathModuleDiscoverer(pluginArtifactClassLoaderToDiscoverModules).discover());
+                MuleModule module = validatePluginModule(pluginUrlClassification.getName(), modules);
+
+                exportedPackages = module.getExportedPackages();
+                exportedResources = module.getExportedPackages();
+            }
+            ArtifactClassLoaderFilter filter = artifactClassLoaderFilterFactory.create(exportedPackages.stream().collect(Collectors.joining(", ")),
+                                                                                       exportedResources.stream().collect(Collectors.joining(", ")));
+            pluginClassLoaders.add(new FilteringArtifactClassLoader(pluginCL, filter));
+        }
+        return new CompositeClassLoader(parent, pluginClassLoaders, childClassLoaderLookupPolicy);
+    }
+
+    /**
+     * Validates that only one module should be discovered.
+     *
+     * @param pluginName
+     * @param discoveredModules
+     * @return the first Module from the list due to there should be only one module.
+     */
+    private MuleModule validatePluginModule(String pluginName, List<MuleModule> discoveredModules)
+    {
+        if (discoveredModules.size() == 0)
+        {
+            throw new IllegalStateException(pluginName + " doesn't have in its classpath a mule-module.properties to define what packages and resources should expose");
+        }
+        if (discoveredModules.size() > 1)
+        {
+            throw new IllegalStateException(pluginName + " has more than one mule-module.properties, composing plugins is not allowed");
+        }
+        return discoveredModules.get(0);
+    }
+
+    /**
+     * Creates an {@link ArtifactClassLoader} for the application.
+     *
+     * @param parent the parent class loader to be assigned to the new one created here
+     * @param childClassLoaderLookupPolicy look policy to be used
+     * @param artifactUrlClassification the url classifications to get plugins urls
+     * @return the {@link ArtifactClassLoader} to be used for test runner as in TCCL.
```

Comment:
The classloader is not just used as the TCCL, it is also used to create the test class. I would say that is the classloader used to run the test